### PR TITLE
Push Rules Into The Schema Layer

### PR DESCRIPTION
**Pull request by fppt**

### CODE REVIEW

**Comment by fppt** [comment]
@aelred Your turn. I also left a super secret TODO for you.

File: grakn-core/src/main/java/ai/grakn/GraknTx.java [hunk 1]
```diff
@@ -116,28 +117,30 @@
...
+     * @param when A string representing the when part of the {@link Rule}
+     * @param then A string representing the then part of the {@link Rule}
+     * @return new or existing {@link Rule} with the provided label.
      *
      * @throws GraknTxOperationException if the graph is closed
-     * @throws PropertyNotUniqueException if the {@param label} is already in use by an existing non-{@link RuleType}.
+     * @throws PropertyNotUniqueException if the {@param label} is already in use by an existing non-{@link Rule}.
      */
-    RuleType putRuleType(String label);
+    Rule putRuleType(String label, Pattern when, Pattern then);
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/GraknTx.java]: This needs to be `putRule`.

File: grakn-core/src/main/java/ai/grakn/GraknTx.java [hunk 2]
```diff
@@ -116,28 +117,30 @@
...
-     * @return new or existing {@link RuleType} with the provided label.
+     * @param label A unique label for the {@link Rule}
+     * @return new or existing {@link Rule} with the provided label.
      *
      * @throws GraknTxOperationException if the graph is closed
-     * @throws PropertyNotUniqueException if the {@param label} is already in use by an existing non-{@link RuleType}.
+     * @throws PropertyNotUniqueException if the {@param label} is already in use by an existing non-{@link Rule}.
      */
-    RuleType putRuleType(Label label);
+    Rule putRuleType(Label label, Pattern when, Pattern then);
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/GraknTx.java]: ditto

File: grakn-core/src/main/java/ai/grakn/GraknTx.java [hunk 3]
```diff
@@ -289,16 +292,16 @@
...
-     * @return The Rule Type with the provided label or null if no such Rule Type exists.
+     * @param label A unique label which identifies the {@link Rule} in the graph.
+     * @return The {@link Rule} with the provided label or null if no such Rule Type exists.
      *
      * @throws GraknTxOperationException if the graph is closed
      */
     @CheckReturnValue
     @Nullable
-    RuleType getRuleType(String label);
+    Rule getRuleType(String label);
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/GraknTx.java]: `getRule`

File: grakn-core/src/main/java/ai/grakn/concept/Concept.java [hunk 4]
```diff
@@ -123,13 +123,13 @@ default RelationshipType asRelationshipType(){
...
-     * Return as a {@link RuleType} if the {@link Concept} is a {@link RuleType}.
+     * Return as a {@link Rule} if the {@link Concept} is a {@link Rule}.
      *
-     * @return A {@link RuleType} if the {@link Concept} is a {@link RuleType}
+     * @return A {@link Rule} if the {@link Concept} is a {@link Rule}
      */
     @CheckReturnValue
-    default RuleType asRuleType(){
-        throw GraknTxOperationException.invalidCasting(this, RuleType.class);
+    default Rule asRuleType(){
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/Concept.java]: `asRule`

File: grakn-core/src/main/java/ai/grakn/concept/Concept.java [hunk 5]
```diff
@@ -242,9 +232,9 @@ default boolean isAttributeType(){
     }
 
     /**
-     * Determine if the {@link Concept} is a {@link RuleType}.
+     * Determine if the {@link Concept} is a {@link Rule}.
      *
-     * @return true if the {@link Concept} is a {@link RuleType}
+     * @return true if the {@link Concept} is a {@link Rule}
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/Concept.java]: `isRule`

File: grakn-core/src/main/java/ai/grakn/concept/Rule.java [hunk 6]
```diff
@@ -21,33 +21,21 @@
 import ai.grakn.graql.Pattern;
 
 import javax.annotation.CheckReturnValue;
+import javax.annotation.Nonnull;
 import java.util.stream.Stream;
 
 /**
  * <p>
- *     A rule which defines how implicit knowledge can extracted.
+ *     An ontological element used to model and categorise different types of {@link Rule}.
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/Rule.java]: This description is wrong. Now these are actually rules themselves. Probably better to take the old description of `Rule` (and update it not to say LHS etc. 😛)

File: grakn-core/src/main/java/ai/grakn/concept/Rule.java [hunk 7]
```diff
@@ -84,34 +71,56 @@
...
+     * Changes the {@link Label} of this {@link Concept} to a new one.
+     * @param label The new {@link Label}.
+     * @return The {@link Concept} itself
+     */
+    Rule setLabel(Label label);
+
     /**
      *
-     * @return The type of this Graql
+     * @return The super type of this {@link Rule}
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/Rule.java]: fun pedantry: we shouldn't say "super type". I've been going with "super concept", but you can just say "super" if you want.

File: grakn-core/src/main/java/ai/grakn/concept/Type.java [hunk 8]
```diff
@@ -38,7 +38,7 @@
  * @see Role
  * @see RelationshipType
  * @see AttributeType
- * @see RuleType
+ * @see Rule
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/Type.java]: This isn't a `Type` so shouldn't be here!

File: grakn-core/src/main/java/ai/grakn/kb/admin/GraknAdmin.java [hunk 9]
```diff
@@ -125,28 +125,12 @@
...
-    RuleType getMetaRuleInference();
-
-    /**
-     * Get the root of all constraint rules.
-     *
-     * @return The meta rule -> constraint-rule.
-     */
-    @CheckReturnValue
-    RuleType getMetaRuleConstraint();
+    Rule getMetaRuleType();
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/kb/admin/GraknAdmin.java]: `getMetaRule`

File: grakn-core/src/main/java/ai/grakn/util/Schema.java [hunk 10]
```diff
@@ -128,13 +125,12 @@ public static boolean isMetaLabel(Label label) {
         RELATIONSHIP_TYPE(RelationshipType.class),
         ATTRIBUTE_TYPE(AttributeType.class),
         ENTITY_TYPE(EntityType.class),
-        RULE_TYPE(RuleType.class),
+        RULE_TYPE(Rule.class),
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/util/Schema.java]: `RULE` I guess? But this is internal so less of an issue.

File: grakn-graql-shell/src/main/java/ai/grakn/graql/GraqlShell.java [hunk 11]
```diff
@@ -123,7 +122,7 @@
     private static final String LICENSE_COMMAND = "license";
     private static final String CLEAN_COMMAND = "clean";
     private static final String HI_POP_COMMAND =
-            HAS.name().substring(0, 1) + INFERENCE_RULE.name().substring(0, 1) +
+            HAS.name().substring(0, 1) + Integer.class.getSimpleName().substring(0, 1) +
```
Comment (aelred) [inline grakn-graql-shell/src/main/java/ai/grakn/graql/GraqlShell.java]:
```
 _,-^-;,-'''''-.
/_  ` )  `      \
`-., _,  ;      /
   )_))_,-_,-/_(
```

File: grakn-kb/src/main/java/ai/grakn/kb/internal/concept/ElementFactory.java [hunk 12]
```diff
@@ -136,13 +135,8 @@ EntityImpl buildEntity(VertexElement vertex, EntityType type){
...
 
-    // ----------------------------------------- Building Rule Types  --------------------------------------------------
-    public RuleTypeImpl buildRuleType(VertexElement vertex, RuleType type){
-        return getOrBuildConcept(vertex, (v) -> new RuleTypeImpl(v, type));
-    }
-
-    // -------------------------------------------- Building Rules
-    RuleImpl buildRule(VertexElement vertex, RuleType type, Pattern when, Pattern then){
+    // ----------------------------------------- Building Rules --------------------------------------------------
+    public RuleImpl buildRuleType(VertexElement vertex, Rule type, Pattern when, Pattern then){
```
Comment (aelred) [inline grakn-kb/src/main/java/ai/grakn/kb/internal/concept/ElementFactory.java]: `buildRule`

File: grakn-migration/base/src/main/java/ai/grakn/migration/base/MigrationCLI.java [hunk 13]
```diff
@@ -136,9 +136,9 @@ public static void printWholeCompletionMessage(MigrationOptions options){
             builder.append("Graph schema contains:\n");
             builder.append("\t ").append(graph.admin().getMetaEntityType().instances().count()).append(" entity types\n");
             builder.append("\t ").append(graph.admin().getMetaRelationType().instances().count()).append(" relation types\n");
-            builder.append("\t ").append("0 role types\n");
```
Comment (aelred) [inline grakn-migration/base/src/main/java/ai/grakn/migration/base/MigrationCLI.java]: haha! is this how we "fixed" this before? 😛

File: grakn-test-tools/src/main/java/ai/grakn/generator/PutSchemaConceptFunctions.java [hunk 14]
```diff
@@ -43,7 +43,6 @@ public PutSchemaConceptFunctions() {
         return random.choose(ImmutableList.of(
                 GraknTx::putEntityType,
                 (graph, label) -> graph.putAttributeType(label, gen(AttributeType.DataType.class)),
-                GraknTx::putRuleType,
```
Comment (aelred) [inline grakn-test-tools/src/main/java/ai/grakn/generator/PutSchemaConceptFunctions.java]: We need to put `GraknTx::putRule` here

File: grakn-test-tools/src/main/java/ai/grakn/generator/Rules.java [hunk 15]
```diff
@@ -19,25 +19,38 @@
...
+    }
+
+    @Override
+    protected Rule metaSchemaConcept() {
+        return tx().admin().getMetaRuleType();
+    }
+
+    @Override
+    protected Collection<Rule> otherMetaSchemaConcepts() {
+        return ImmutableSet.of(tx().admin().getMetaRuleType());
```
Comment (aelred) [inline grakn-test-tools/src/main/java/ai/grakn/generator/Rules.java]: I think we don't need to override this method now.

File: grakn-test/test-integration/src/test/java/ai/grakn/test/property/ConceptPropertyTest.java [hunk 16]
```diff
@@ -249,14 +248,6 @@ public void whenConceptIsNotAResource_TheConceptCannotBeConvertedToAResource(Con
         concept.asAttribute();
     }
 
-    @Property
-    public void whenConceptIsNotARule_TheConceptCannotBeConvertedToARule(Concept concept) {
-        assumeFalse(concept.isRule());
-        exception.expect(GraknTxOperationException.class);
-        //noinspection ResultOfMethodCallIgnored
-        concept.asRule();
-    }
```
Comment (aelred) [inline grakn-test/test-integration/src/test/java/ai/grakn/test/property/ConceptPropertyTest.java]: This test (and the one above) are still relevant right?

File: grakn-test/test-integration/src/test/java/ai/grakn/test/property/GraknTxPutPropertyTest.java [hunk 17]
```diff
@@ -186,21 +185,21 @@ public void whenCallingPutResourceTypeWithAnExistingNonUniqueResourceTypeLabelBu
     }
 
     @Property
-    public void whenCallingPutRuleType_CreateATypeWithSuperTypeRule(@Open GraknTx graph, @Unused Label label) {
-        RuleType ruleType = graph.putRuleType(label);
-        assertEquals(graph.admin().getMetaRuleType(), ruleType.sup());
+    public void whenCallingPutRuleType_CreateATypeWithSuperTypeRule(@Open GraknTx tx, @Unused Label label) {
```
Comment (aelred) [inline grakn-test/test-integration/src/test/java/ai/grakn/test/property/GraknTxPutPropertyTest.java]: `PutRule` and so on

File: grakn-core/src/main/java/ai/grakn/GraknTx.java (hunk 1 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/GraknTx.java]: Good catch!

File: grakn-core/src/main/java/ai/grakn/GraknTx.java (hunk 3 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/GraknTx.java]: Yar!

File: grakn-core/src/main/java/ai/grakn/concept/Concept.java (hunk 4 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/concept/Concept.java]: 🍌

File: grakn-core/src/main/java/ai/grakn/concept/Concept.java (hunk 5 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/concept/Concept.java]: 🍌

File: grakn-core/src/main/java/ai/grakn/concept/Rule.java (hunk 6 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/concept/Rule.java]: 🍌

File: grakn-core/src/main/java/ai/grakn/concept/Rule.java (hunk 7 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/concept/Rule.java]: 🍌

File: grakn-core/src/main/java/ai/grakn/concept/Type.java (hunk 8 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/concept/Type.java]: 🍌

File: grakn-core/src/main/java/ai/grakn/kb/admin/GraknAdmin.java (hunk 9 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/kb/admin/GraknAdmin.java]: 🍌

File: grakn-core/src/main/java/ai/grakn/util/Schema.java (hunk 10 above)
Comment (fppt) [reply grakn-core/src/main/java/ai/grakn/util/Schema.java]: 🍌

File: grakn-graql-shell/src/main/java/ai/grakn/graql/GraqlShell.java (hunk 11 above)
Comment (fppt) [reply grakn-graql-shell/src/main/java/ai/grakn/graql/GraqlShell.java]: 🐘

File: grakn-kb/src/main/java/ai/grakn/kb/internal/concept/ElementFactory.java (hunk 12 above)
Comment (fppt) [reply grakn-kb/src/main/java/ai/grakn/kb/internal/concept/ElementFactory.java]: 🍌

File: grakn-migration/base/src/main/java/ai/grakn/migration/base/MigrationCLI.java (hunk 13 above)
Comment (fppt) [reply grakn-migration/base/src/main/java/ai/grakn/migration/base/MigrationCLI.java]: 🍌

File: grakn-test-tools/src/main/java/ai/grakn/generator/PutSchemaConceptFunctions.java (hunk 14 above)
Comment (fppt) [reply grakn-test-tools/src/main/java/ai/grakn/generator/PutSchemaConceptFunctions.java]: 🍌

**Pull request merged**